### PR TITLE
commands,lfs: handle malformed pointers

### LINKED ISSUE #1729: Decoding a pointer is a fatal error - should we just warn?
If GitLFS cannot decode a pointer file then it fails. This can happen if a user tracks `*.bin` files on a case *sensitive* file system. GitLFS would not track `*.BIN` files on that system. If another users works with the repo on a case *insensitive* file system then GitLFS would try to decode pointers for the `*.BIN` which would fail.  I wonder if it should rather just warn the user in that case? 

Plus the failure is different for large and small files (@ttaylorr I assume this is because we read the first 1024 bytes of the pointer, right?).

## Testcase:

```
git version 2.11.0
git-lfs/1.5.2 (GitHub; darwin amd64; go 1.7.3)
```

Error message if decoding of a large file fails:
```
$ git clone --branch largefile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 33, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 33 (delta 2), reused 0 (delta 0), pack-reused 21
Receiving objects: 100% (33/33), 6.21 KiB | 0 bytes/s, done.
Resolving deltas: 100% (5/5), done.
fatal: big-file1.bin: smudge filter lfs failed
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x3a1400, 0xc420010250)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).readRequest(0xc420172fc0, 0xc4200f51f0, 0x0, 0xc420051a10)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:181 +0x218
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).Scan(0xc420172fc0, 0x402e11)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:143 +0x50
github.com/git-lfs/git-lfs/commands.filterCommand(0xc4200e9380, 0x602ba0, 0x0, 0x0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/command_filter_process.go:76 +0x1f9
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200e9380, 0x602ba0, 0x0, 0x0, 0xc4200e9380, 0x602ba0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:477 +0x170
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200e84e0, 0xc420051ed0, 0x1)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:551 +0x380
github.com/git-lfs/git-lfs/commands.Run()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/run.go:65 +0x196
main.main()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/git-lfs.go:33 +0x111
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

Error message if decoding of a small file fails:
```
$ git clone --branch smallfile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 33, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 33 (delta 2), reused 0 (delta 0), pack-reused 21
Receiving objects: 100% (33/33), 6.21 KiB | 0 bytes/s, done.
Resolving deltas: 100% (5/5), done.
fatal: big-file1.bin: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

Plus, even if we disable the smudge filter, the checkout fails:
```
$ GIT_LFS_SKIP_SMUDGE=1 git clone --branch largefile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 33, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 33 (delta 2), reused 0 (delta 0), pack-reused 21
Receiving objects: 100% (33/33), 6.21 KiB | 0 bytes/s, done.
Resolving deltas: 100% (5/5), done.
fatal: big-file1.bin: smudge filter lfs failed
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x3a1400, 0xc420010250)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).readRequest(0xc420135dc0, 0xc420106a30, 0x0, 0xc420051a10)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:181 +0x218
github.com/git-lfs/git-lfs/git.(*FilterProcessScanner).Scan(0xc420135dc0, 0x402e11)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/git/filter_process_scanner.go:143 +0x50
github.com/git-lfs/git-lfs/commands.filterCommand(0xc4200f8000, 0x602ba0, 0x0, 0x0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/command_filter_process.go:76 +0x1f9
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200f8000, 0x602ba0, 0x0, 0x0, 0xc4200f8000, 0x602ba0)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:477 +0x170
github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42008d040, 0xc420051ed0, 0x1)
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/vendor/github.com/spf13/cobra/command.go:551 +0x380
github.com/git-lfs/git-lfs/commands.Run()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/src/github.com/git-lfs/git-lfs/commands/run.go:65 +0x196
main.main()
	/private/tmp/git-lfs-20161122-73173-1n7qhbi/git-lfs-1.5.2/git-lfs.go:33 +0x111
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

### PR DESCRIPTION
**Pull request by ttaylorr**
This pull-request fixes #1729 by streaming out the entire contents of a malformed pointer, instead of truncating the first 1024 bytes.

Previously, if LFS was unable to parse file that it thought was a pointer, it would "smudge" the file by just emitting it as is. The catch, however, was that the only data that it could hold onto was the first 1024 bytes. Since `(*git.PktlineReader) Read()`s block if there is no data on the pipe, buffering a fixed amount of data was the best that we could do. There were two cases:

1. `len(pbuf) < 1024`: We received the entire contents, no join necessary.
2. `len(pbuf) == 1024`: We _might_ have received the entire contents, or there were more than 1024 bytes and we just read the first 1024 of them. 

To handle this better, I introduced `lfs.DecodePointerHasMore` which mimics `lfs.DecodePointer`, but returns an additional `bool`, dictating whether or not more data is available to be read (in other words, where the `(*git.PktlineReader) Read()` error was `== io.EOF` or not). When using this in the `commands` package, we just check if there was `more` data, and join the reader as necessary:

```go
var r io.Reader = bytes.NewBuffer(pbuf)
if more {
        r = io.MultiReader(r, src)
}
```

As an added bonus, I added some warning messages that get printed to `os.Stderr` and include the filenames of all malformed pointers. When using `filter.lfs.process`, the files are collected together and printed once at the end. When using the `filter.lfs.clean` mode, we print out an error once per file, instead of bundling them up. Short of storing the list on disk in `.git/`, this is the best that we can do.

To wrap things up, I audited all uses of `lfs.DecodePointer`, `lfs.DecodeFromFile`, et. al., and determined that there were no other instances in the code where we would only output partially buffered data.

Hooray!

---

Here's a quick break-down of what was done:

1. cfcba6f: Introduce `lfs.DecodePointerHasMore` and use it in `commands.smudge()` to determine when to correctly join and copy the readers.
2. b3bab24: Keep track of the malformed pointers seen in a given pull/checkout/clone and emit them at the end (or once per file, if we're using `filter.lfs.clean` instead of the process filter).

---

/cc @git-lfs/core #1729

### CODE REVIEW

File: commands/command_filter_process.go [hunk 1]
```diff
@@ -89,11 +63,16 @@ func filterCommand(cmd *cobra.Command, args []string) {
 			err = clean(w, req.Payload, req.Header["pathname"])
 		case "smudge":
 			w = git.NewPktlineWriter(os.Stdout, smudgeFilterBufferCapacity)
-			err = filterSmudge(w, req.Payload, req.Header["pathname"], skip, filter)
+			err = smudge(w, req.Payload, req.Header["pathname"], skip, filter)
```
Comment (ttaylorr) [inline commands/command_filter_process.go]: As an added benefit, I got rid of the intermediate `filterSmudge` function, so now the legacy `filter.lfs.{clean,smudge}` and `filter.lfs.process` share the _exact same_ implementation! ✨

File: lfs/pointer.go [hunk 2]
```diff
@@ -109,16 +109,28 @@ func DecodePointer(reader io.Reader) (*Pointer, error) {
...
+}
+
+// DecodeFromHasMore decodes an *lfs.Pointer from the given io.Reader, "reader".
+// If the pointer encoded in the reader could successfully be read and decoded,
+// it will be returned with a nil error.
+//
+// If the pointer could not be decoded, a buffer of the data read so far, along
+// with whether or not there was more data not yet consumed will be returned
+// instead.
+func DecodeFromHasMore(reader io.Reader) ([]byte, bool, *Pointer, error) {
```
Comment (technoweenie) [inline lfs/pointer.go]: Just for posterity, I mentioned in video chat that the `DecodeFromHasMore` return signature feels clunky. My suggestion was to change the `DecodeFrom` interface to `(io.Reader, *Pointer, error)`, and let the caller just drain the returned reader if it wants. Callers shouldn't have to worry about how to join the read buffer with the io reader.

**Comment by larsxschneider** [comment]
Thanks @ttaylorr ! This works great. One minor nit. The `Encountered 1 file(s) that should have been pointers, but weren't:` does not work with `git lfs clone`. Consider this:

```
$ git clone --branch largefile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 40, done.
remote: Total 40 (delta 0), reused 0 (delta 0), pack-reused 40
Receiving objects: 100% (40/40), 6.75 KiB | 0 bytes/s, done.
Resolving deltas: 100% (10/10), done.
Encountered 1 file(s) that should have been pointers, but weren't:
    big-file1.bin

$ git lfs clone --branch largefile git@github.com:larsxschneider/lfstest-individual-files.git smudge-error
Cloning into 'smudge-error'...
remote: Counting objects: 40, done.
remote: Total 40 (delta 0), reused 0 (delta 0), pack-reused 40
Receiving objects: 100% (40/40), 6.75 KiB | 0 bytes/s, done.
Resolving deltas: 100% (10/10), done.
```

**Comment by ttaylorr** [comment]
> One minor nit. The `Encountered 1 file(s) that should have been pointers, but weren't:` does not work with `git lfs clone`.

@larsxschneider: thanks for pointing this out! If my memory serves, the LFS `clone` command uses `commands.pull()` to do the checkout, which receives `*lfs.WrappedPointer`s from the `*lfs.gitscanner` [here][1]. I think in order to make this work I'd just need to make `lfs.DecodePointer` return an error, or teach the gitscanner how to emit the invalid pointer errors.

I'll open an issue.

[1]: https://github.com/git-lfs/git-lfs/blob/85f581166bba18dce6484e34ab5ba6dc03387405/commands/command_pull.go#L51

**Comment by larsxschneider** [comment]
Unfortunately I found a bug if the invalid files exceed a certain size threshold.
The following clone will hang [here](https://github.com/git-lfs/git-lfs/blob/a0d9487e64d8059410fe9d4aa83c1cfb06477002/commands/command_smudge.go#L33):

```
git clone --branch largefile2 git@github.com:larsxschneider/lfstest-individual-files.git smudge-error2
```

I tested it on `git version 2.11.0` with `git-lfs/1.5.4 (GitHub; darwin amd64; go 1.7.4)` and `git lfs lastest master: a0d9487e`.

**Comment by ttaylorr** [comment]
I can definitely reproduce this when cloning that same repository myself. What's
curious to me is that this seems unrelated to the `blobSizeCutoff` const, since
we have integration tests for 1023, 1024 and 1025-byte malformed pointers.

Perhaps this has to do with the implementation of `io.Copy`, and buffering
between Git<->LFS.

@larsxschneider - do you know of a particular filesize (or range) that causes
this behavior to occur?

**Pull request merged**